packages/zoe/src/contracts/multipoolAutoswap/pool.js

```diff
@@ -0,0 +1,175 @@
+            liqTokenSupply,
+            inputValue: userSeat.getAmountAllocated('Central').value,
+            inputReserve: pool.getCentralAmount().value,
+          }),
+        );
+
+        const liquidityAmountOut = liquidityAmountMath.make(liquidityValueOut);
+        liquidityZcfMint.mintGains({ Liquidity: liquidityAmountOut }, poolSeat);
+        liqTokenSupply += liquidityValueOut;
+
+        trade(
+          zcf,
+          {
+            seat: poolSeat,
+            gains: userSeat.getCurrentAllocation(),
+          },
+          {
+            seat: userSeat,
+            gains: { Liquidity: liquidityAmountOut },
+          },
+        );
+        userSeat.exit();
+        return 'Added liquidity.';
+      },
+      removeLiquidity: userSeat => {
+        const liquidityIn = userSeat.getAmountAllocated(
+          'Liquidity',
+          liquidityBrand,
+        );
+        const liquidityValueIn = liquidityIn.value;
+        const centralTokenAmountOut = pool.getCentralAmountMath().make(
+          calcValueToRemove(
+            harden({
+              liqTokenSupply,
+              poolValue: pool.getCentralAmount().value,
+              liquidityValueIn,
+            }),
+          ),
+        );
+
+        const tokenKeywordAmountOut = pool.getAmountMath().make(
+          calcValueToRemove(
+            harden({
+              liqTokenSupply,
+              poolValue: pool.getSecondaryAmount().value,
+              liquidityValueIn,
+            }),
+          ),
+        );
+
+        liqTokenSupply -= liquidityValueIn;
+
+        trade(
+          zcf,
+          {
+            seat: poolSeat,
+            gains: { Liquidity: liquidityIn },
+          },
+          {
+            seat: userSeat,
+            gains: {
+              Central: centralTokenAmountOut,
+              Secondary: tokenKeywordAmountOut,
+            },
+          },
+        );
+
+        userSeat.exit();
+        return 'Liquidity successfully removed.';
+      },
+    };
+    return pool;
+  };
+
+  /**
+   * Allows users to add new liquidity pools. `secondaryIssuer` and
+   * its keyword must not have been already used
+   * @param {Issuer} secondaryIssuer
+   * @param {Keyword} keyword - will be used in the
+   * terms.issuers for the contract, but not used otherwise
+   */
+  const addPool = async (secondaryIssuer, keyword) => {
+    zcf.assertUniqueKeyword(keyword);
+    const liquidityKeyword = `${keyword}Liquidity`;
+    zcf.assertUniqueKeyword(liquidityKeyword);
+    const secondaryBrand = await E(secondaryIssuer).getBrand();
+    const brandMatches = await E(secondaryBrand).isMyIssuer(secondaryIssuer);
+    assert(
+      brandMatches,
+      `The provided issuer was using another issuer's brand`,
+    );
+    assert(
+      !isSecondary(secondaryBrand),
+      details`issuer ${secondaryIssuer} already has a pool`,
+    );
+
+    await zcf.saveIssuer(secondaryIssuer, keyword);
+    assertUsesNatMath(zcf, secondaryBrand);
```

Comment:
check for validity before saving the issuer.

Reply:
What validity checks aren't being done?

Reply:
I meant that `assertUsesNatMath()` is invoked after saving the issuer. It should come before.